### PR TITLE
Create Reactive providers via `rx` or `reactive`

### LINKED ISSUE #1144: Update ReactiveCocoa documentation to ReactiveSwift
The documentation still refers to `ReactiveSwift` as `ReactiveCocoa` in both the examples' [README](https://github.com/Moya/Moya/blob/master/docs/Examples/README.md) as well as the [documentation itself](https://github.com/Moya/Moya/blob/master/docs/Examples/ReactiveCocoa.md) (including the document's name).

This should be updated to use the `ReactiveSwift` name. Make sure to also fix all references to these documents - it seems like the only reference to it is in the README: https://github.com/Moya/Moya/search?l=Markdown&q=reactivecocoa.md&type=&utf8=✓.

### PR DESCRIPTION
**Pull request by BasThomas**
Supersedes #963. (cc @kemchenj)

Also fixes #1144.

### CODE REVIEW

File: Sources/ReactiveMoya/ReactiveSwiftMoyaProvider.swift [hunk 1]
```diff
@@ -17,6 +18,11 @@ open class ReactiveSwiftMoyaProvider<Target>: MoyaProvider<Target> where Target:
         self.stubScheduler = stubScheduler
         super.init(endpointClosure: endpointClosure, requestClosure: requestClosure, stubClosure: stubClosure, manager: manager, plugins: plugins, trackInflights: trackInflights)
     }
+    
+    fileprivate init(moyaProvider: MoyaProvider<Target>) {
+        self.stubScheduler = nil
```
Comment (BasThomas) [inline Sources/ReactiveMoya/ReactiveSwiftMoyaProvider.swift]: Not quite sure what to do here - right now, using `.reactive` means the `stubScheduler` will always be `nil`. Should this then be a `var` instead of a `let`, or can we solve this in another way?

**Comment by BasThomas** [comment]
Targeted this at the `master` branch by accident, changing the base branch to `9.0.0-dev` adds a bunch of commits. 🤔

**Comment by BasThomas** [comment]
Also, from #963:

> Then let `MoyaProvider` conform to `ReactiveCompatible` and `ReactiveExtensionsProvider`, then we could change our code like this:

It seems like we can't use both of these protocols, seeing how they are implemented ([ReactiveCompatible](https://github.com/ReactiveX/RxSwift/blob/0b66f666ba6955a51cba1ad530311b030fa4db9c/RxSwift/Reactive.swift#L38); [ReactiveExtensionsProvider](https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Sources/Reactive.swift#L9)). Or at least, I don't really see how to do so. Am I missing something there?

File: Sources/ReactiveMoya/ReactiveSwiftMoyaProvider.swift (hunk 1 above)
Comment (sunshinejr) [inline Sources/ReactiveMoya/ReactiveSwiftMoyaProvider.swift]: Hmm, because of the queues we're adding in the other PR, I think this scheduler could be removed and then making this extensions would be quite easy.

File: Sources/RxMoya/RxMoyaProvider.swift [hunk 2]
```diff
@@ -69,3 +74,10 @@ public extension RxMoyaProvider {
         }
     }
 }
+
+public extension MoyaProvider {
+    
+    public var rx: RxMoyaProvider<Target> {
```
Comment (sunshinejr) [inline Sources/RxMoya/RxMoyaProvider.swift]: In Rx to make something in the `rx` prefix, one would usually do:
```swift
extension Reactive where Base: MoyaProvider {
    func request() -> Observable<Response> {...}
}
```

It is probably similar in ReactiveSwift.

**Comment by BasThomas** [comment]
Interesting CI failure:

```
❌  /Users/distiller/Moya/Tests/ReactiveSwiftMoyaProviderTests.swift:190:92: argument 'stubScheduler' must precede argument 'stubClosure'

                provider = MoyaProvider<GitHub>(stubClosure: MoyaProvider.immediatelyStub, stubScheduler: testScheduler).reactive
                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It seems like the parameter order differs from `ReactiveSwiftMoyaProvider` and `MoyaProvider`. 😯

**Comment by sunshinejr** [comment]
@BasThomas you probably just need to extend `Reactive` (`RxSwift`.`Reactive` & `ReactiveSwift`.`Reactive`, look at my comment in your commit) and then we can skip using other providers and provide `request` method.

**Comment by BasThomas** [comment]
@sunshinejr I've been trying to update the tests and verify if everything is working locally, but can't seem to get the tests running. For this I think I need to do a `swift build` and `swift test`, right? Problem is, `swift build` errors:

```bash
bas$ swift build
warning: refname '4.5.0' is ambiguous.
warning: refname '4.5.0' is ambiguous.
HEAD is now at 7fe7b91 Added release notes to the CHANGELOG and bumped the version to 4.5.0.
Resolved version: 4.5.0
warning: refname '1.1.3' is ambiguous.
warning: refname '1.1.3' is ambiguous.
HEAD is now at 92d73573 1.1.3
Resolved version: 1.1.3
warning: refname '3.2.3' is ambiguous.
warning: refname '3.2.3' is ambiguous.
HEAD is now at c844618 3.2.3
Resolved version: 3.2.3
/Users/bas/Documents/Programming/iOS/Moya/Packages/RxSwift/Package.swift:5:36: error: use of unresolved identifier 'ProcessInfo'
let RxTestIsTarget = buildTests || ProcessInfo.processInfo.environment["TEST"] == "1"
                                   ^~~~~~~~~~~
Foundation.NSProcessInfo:1:14: note: did you mean 'NSProcessInfo'?
public class NSProcessInfo : NSObject {
             ^
ApplicationServices.ProcessInfoRec:1:15: note: did you mean 'ProcessInfoRec'?
public struct ProcessInfoRec {
              ^
error: exit(1): /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-05-31-a.xctoolchain/usr/bin/swiftc --driver-mode=swift -I /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-05-31-a.xctoolchain/usr/lib/swift/pm -L /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-05-31-a.xctoolchain/usr/lib/swift/pm -lPackageDescription -target x86_64-apple-macosx10.10 /Users/bas/Documents/Programming/iOS/Moya/Packages/RxSwift/Package.swift -fileno 3
```

Any idea?

**Comment by sunshinejr** [comment]
@BasThomas We don't support `swift test` yet, it is on my todo list to make it work some day and there is an issue for that as well (#894). Right now we can use Carthage for it (`carthage update --platform iOS` should be fine) and then open Moya.xcodeproj and run the tests in. Should work fine, but if you encounter any issue let me know 👍

**Comment by BasThomas** [comment]
Interesting, I had two *really* old snapshots, though Xcode was just using the one bundled with Xcode; I wonder how it used one of those other snapshots when running `swift build`. 🤷‍♂️

**Comment by BasThomas** [comment]
Testing via Carthage works; I've created #1141 to make a short `.md` to document this.

**Comment by BasThomas** [comment]
Added a DNM, as we would need to merge this into `9.0.0-dev`, not `master`.

File: Sources/RxMoya/RxMoyaProvider.swift (hunk 2 above)
Comment (BasThomas) [inline Sources/RxMoya/RxMoyaProvider.swift]: I am not quite sure what this does / how that would work? Would this drop `RxMoyaProvider` completely, add the `.rx` var and make only `request()` be enabled for the returned value of `.rx`?

Comment (sunshinejr) [inline Sources/RxMoya/RxMoyaProvider.swift]: I think we just need to conform to `ReactiveCompatible` like:
```swift
extension MoyaProvider: ReactiveCompatible {}
```

And the same for ReactiveSwift. This way we get `.rx` & `.reactive` properties for free and we could remove `RxProvider` & `ReactiveProvider`, using only `rx.request` & `reactive.request` (which we have already implemented, just need to extract request functions from providers into extensions to `Reactive` structs)

**Comment by AndrewSB** [comment]
woah, I just switched the base from master to 9.0.0-dev, and it looks like 9.0.0-dev is in dire need of a rebase :octocat:

**Comment by BasThomas** [comment]
Yeah, that's what I saw when switching the target branch as well. 🤔

**Comment by AndrewSB** [comment]
somebody's who's familiar with the recent commits on master should run
```
git checkout 9.0.0-dev
git rebase master
git push 9.0.0-dev
```

and the unrelated commits should disappear from this PR

**Comment by sunshinejr** [comment]
I did the rebase on 9.0.0-dev and fixed conflicts on this branch, should be alright. In case something is wrong, I've got the copy of the 9.0.0-dev branch so it's all good 👍

**Comment by sunshinejr** [comment]
And how are we with that extension change to `ReactiveCompatible`s, @BasThomas? What do you think? Kinda got buried in the process of this PR, so it is hard to follow the conversation :D

**Comment by BasThomas** [comment]
Will take a look!

**Comment by BasThomas** [comment]
I've just updated to use `ReactiveExtensionsProvider` and `ReactiveCompatible`. The build is expected to fail - I haven't updated all tests yet, and the extensions don't implement the `request()`. method yet.

Does anyone else want to look into that? As I don't have too much experience with Reactive, I feel some of you might be better for that job. :)

**Comment by BasThomas** [comment]
Also, might both these extensions creating a type of `Reactive<MoyaProvider<Target>>` cause a problem somewhere? 🤔

**Comment by sunshinejr** [comment]
@BasThomas I might do one commit that will show what I had in mind, what do you think? In case something went wrong we can just remove it :D

**Comment by BasThomas** [comment]
Definitely, go for it!

**Comment by BasThomas** [comment]
I've done another update to the docs, meanwhile fixing #1144 as well.

**Comment by sunshinejr** [comment]
I did some work on that and it seems like we would need to add a protocol for `MoyaProvider` to make extensions happen, because we can't still do something like: 
```swift
extension Reactive where MoyaProvider {}
```

And also we would need to get rid of overriding `stubRequest` in `ReactiveSwift` provider - but I think that we established that queues are gonna implement it. I just made a comment on queues PR for that to make sure we do not forget about it. 

Due to it being a bigger change than 1 commit + a lot of discussion happened there, I think we should do another PR with clear state. What do you think, @BasThomas?

**Comment by BasThomas** [comment]
Yeah, sure!

File: docs/Examples/ReactiveSwift.md [hunk 3]
```diff
@@ -1,10 +1,10 @@
-# ReactiveCocoa examples
+# ReactiveSwift examples
 
-A `ReactiveCocoaMoyaProvider` can be created much like a
+A `ReactiveSwift`-based `MoyaProvider` can be created much like a
 [`MoyaProvider`](../Providers.md) and can be used as follows:
 
 ```swift
-let GitHubProvider = ReactiveCocoaMoyaProvider<GitHub>()
+let gitHubProvider = MoyaProvider<GitHub>().reactive
```
Comment (SD10) [inline docs/Examples/ReactiveSwift.md]: Nit suggestion but we name this `gitHubProvider` then access `provider` in the example.

Comment (BasThomas) [inline docs/Examples/ReactiveSwift.md]: Good catch, will fix both :)

**Comment by pedrovereza** [comment]
@BasThomas @sunshinejr Should this closed in favor of #1153? 🤔

**Comment by BasThomas** [comment]
Yep!
